stats/opencensus/e2e_test.go

```diff
@@ -0,0 +1,773 @@
+/*
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package opencensus
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/internal/grpctest"
+	"google.golang.org/grpc/internal/leakcheck"
+	"google.golang.org/grpc/internal/stubserver"
+	"google.golang.org/grpc/test/grpc_testing"
+)
+
+type s struct {
+	grpctest.Tester
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+func init() {
+	// OpenCensus, once included in binary, will spawn a global goroutine
+	// recorder that is not controllable by application.
+	// https://github.com/census-instrumentation/opencensus-go/issues/1191
+	leakcheck.RegisterIgnoreGoroutine("go.opencensus.io/stats/view.(*worker).start")
+}
+
+var defaultTestTimeout = 5 * time.Second
+
+type fakeExporter struct {
+	t *testing.T
+
+	mu        sync.RWMutex
+	seenViews map[string]*viewInformation
+}
+
+// viewInformation is information Exported from the view package through
+// ExportView relevant to testing, i.e. a reasonably non flaky expectation of
+// desired emissions to Exporter.
+type viewInformation struct {
+	aggType    view.AggType
+	aggBuckets []float64
+	desc       string
+	tagKeys    []tag.Key
+	rows       []*view.Row
+}
+
+func (fe *fakeExporter) ExportView(vd *view.Data) {
+	fe.mu.Lock()
+	defer fe.mu.Unlock()
+	fe.seenViews[vd.View.Name] = &viewInformation{
+		aggType:    vd.View.Aggregation.Type,
+		aggBuckets: vd.View.Aggregation.Buckets,
+		desc:       vd.View.Description,
+		tagKeys:    vd.View.TagKeys,
+		rows:       vd.Rows,
+	}
+}
+
+func (vi *viewInformation) equal(vi2 *viewInformation) bool {
+	if vi == nil && vi2 == nil {
+		return true
+	}
+	if (vi != nil) != (vi2 != nil) {
```

Comment:
Optional, nit: `if vi == nil || vi2 == nil { return false }` is equivalent and will save a comp and allow short circuiting sometimes.